### DISCUSSION TITLE
About listeners

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#241] [General]
Seeing this and not fully understanding it. I want to have a listener on a backend app. Can I have the listener active "forever"?

```dart
@UseResult('Listener will leak in the container if it is not disposed')
  void Function() listen(CapsuleListener listener) {
    // Create a temporary *non-idempotent* capsule so that it doesn't get
    // idempotent garbage collected
    void capsule(CapsuleHandle use) {
      use.asListener();
      listener(use);
    }

    // Put the temporary capsule into the container so it gets data updates
    final manager = _managerOf(capsule);

    return manager.dispose;
  }
```

**Comment by busslina** [comment]
Sorry, didn't read the following, so I will do as it's said there:

```
If you want to listen to capsule(s) *not temporarily*, instead just make
a non-idempotent capsule and [read] it once to initialize it.
(See the documentation for more on this topic.)
```

**Comment by busslina** [comment]
BTW, after a long time I dind't use `container.listen` I crash today with an old error: trying to modify capsules from inside the listener.

The solution is easy: To add an async delay before that, but I don't know if you advice of this issue in the docs. If not I encourage you to do so, as it could prevent the same issue from other `rearch` users. I just did a test repo I share here with very little code:

https://github.com/busslina/2024-12-14-rearch-issue

```dart
import 'package:rearch/rearch.dart';

final container = CapsuleContainer();

void main(List<String> arguments) {
  print('Dart Issue -- Start');

  // ignore: unused_result
  container.listen((use) {
    final appState = use(appStateCapsule);

    print('App state --> ${appState.value}');

    Future.microtask(() {
      switch (appState.value) {
        case AppState.one:
          appState.value = AppState.two;

        case AppState.two:
          appState.value = AppState.three;

        case AppState.three:
          print('Final state --> 3');
      }
    });
  });

  print('Dart Issue -- End');
}

enum AppState {
  one,
  two,
  three,
}

ValueWrapper<AppState> appStateCapsule(CapsuleHandle use) =>
    use.data(AppState.one);
```

**Reply by busslina** [reply]
This applies too "Capsule Listeners". I updated my test repo to include both cases:
- https://github.com/busslina/2024-12-14-rearch-issue/blob/master/bin/rearch_issue.container_listener.dart
- https://github.com/busslina/2024-12-14-rearch-issue/blob/master/bin/rearch_issue.capsule_listener.dart

**Comment by busslina** [comment]
Also you say:
```
If you want to listen to capsule(s) *not temporarily*, instead just make
a non-idempotent capsule and [read] it once to initialize it.
(See the documentation for more on this topic.)
```

The issue here is that using `container.listen` it works great but using instead a "listener" capsule and reading it once is not enough. It seems I have to constantly read it to have the same behavior as with `container.listen`.

Am I right? Or maybe I'm missing something?

Thanks.

**Reply by GregoryConrad** [reply]
You’re missing a `use.asListener();` in the listener capsule :)
